harper-core/src/linting/punctuation_clusters.rs

```diff
@@ -0,0 +1,239 @@
+            TokenKind::Punctuation(Punctuation::Period) => '.',
+            TokenKind::Punctuation(Punctuation::Ampersand) => '&',
+            _ => unreachable!("`char_of` called on non-candidate punctuation"),
+        }
+    }
+}
+
+impl Linter for PunctuationClusters {
+    fn lint(&mut self, document: &Document) -> Vec<Lint> {
+        let toks = document.get_tokens();
+        let mut lints = Vec::new();
+        let mut i = 0;
+
+        while i < toks.len() {
+            // Start of a potential cluster
+            if !Self::is_candidate(&toks[i].kind) {
+                i += 1;
+                continue;
+            }
+
+            let start = i;
+            let mut end = i;
+            let mut uniq = HashSet::<char>::new();
+
+            // Consume the cluster (allowing spaces/newlines in between)
+            while i < toks.len() {
+                match &toks[i].kind {
+                    k if Self::is_candidate(k) => {
+                        uniq.insert(Self::char_of(k));
+                        end = i;
+                        i += 1;
+                    }
+                    TokenKind::Space(_) | TokenKind::Newline(_) => {
+                        end = i;
+                        i += 1;
+                    }
+                    _ => break,
+                }
+            }
+
+            // How many candidate tokens were there?
+            let count = (start..=end)
+                .filter(|idx| Self::is_candidate(&toks[*idx].kind))
+                .count();
+
+            if count >= 2 {
+                let span = Span::new(toks[start].span.start, toks[end].span.end);
+
+                // One suggestion per distinct glyph in the cluster
+                let suggestions = uniq
+                    .into_iter()
+                    .map(|c| Suggestion::ReplaceWith(vec![c]))
+                    .collect::<Vec<_>>();
+
+                lints.push(Lint {
+                    span,
+                    lint_kind: LintKind::Formatting,
+                    suggestions,
+                    message: "Condense this punctuation cluster to a single mark.".into(),
+                    priority: 63,
+                });
+            }
+        }
+
+        lints
+    }
+
+    fn description(&self) -> &str {
+        "Detects consecutive or mixed punctuation marks that should be reduced \
+         to a single comma, semicolon, colon, slash, question mark, \
+         exclamation mark, period, or ampersand."
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::linting::tests::assert_lint_count;
+
+    use super::PunctuationClusters;
+
+    #[test]
+    fn flags_double_comma() {
+        assert_lint_count("Wait,, what happened?", PunctuationClusters, 1);
+    }
+
+    #[test]
+    fn flags_double_semicolon() {
+        assert_lint_count("He hesitated;; then spoke.", PunctuationClusters, 1);
+    }
+
+    #[test]
+    fn flags_double_colon() {
+        assert_lint_count("Choices:: A or B.", PunctuationClusters, 1);
```

Comment:
This could match some language notation for classes. I don't know if it shouldn't be in another rule

Reply:
Methods and namespaces in various languages.